tests/integration/feature_repos/test_repo_configuration.py

```diff
@@ -0,0 +1,343 @@
+        if "FileDataSourceCreator" in c.offline_store_creator:
+            new_configs.append(c)
+        elif "RedshiftDataSourceCreator" in c.offline_store_creator:
+            for p in ["local", "aws"]:
+                new_configs.append(replace(c, provider=p))
+        elif "BigQueryDataSourceCreator" in c.offline_store_creator:
+            for p in ["local", "gcp"]:
+                new_configs.append(replace(c, provider=p))
+    return new_configs
+
+
+@contextmanager
+def construct_test_environment(
+    test_repo_config: TestRepoConfig,
+    create_and_apply: bool = False,
+    materialize: bool = False,
+) -> Environment:
+    """
+    This method should take in the parameters from the test repo config and created a feature repo, apply it,
+    and return the constructed feature store object to callers.
+
+    This feature store object can be interacted for the purposes of tests.
+    The user is *not* expected to perform any clean up actions.
+
+    :param test_repo_config: configuration
+    :return: A feature store built using the supplied configuration.
+    """
+    df = create_dataset()
+
+    project = f"test_correctness_{str(uuid.uuid4()).replace('-', '')[:8]}"
+
+    module_name, config_class_name = test_repo_config.offline_store_creator.rsplit(
+        ".", 1
+    )
+
+    offline_creator: DataSourceCreator = importer.get_class_from_type(
+        module_name, config_class_name, "DataSourceCreator"
+    )(project)
+    ds = offline_creator.create_data_source(
+        project, df, field_mapping={"ts_1": "ts", "id": "driver_id"}
+    )
+    offline_store = offline_creator.create_offline_store_config()
+    online_store = test_repo_config.online_store
+
+    with tempfile.TemporaryDirectory() as repo_dir_name:
+        config = RepoConfig(
+            registry=str(Path(repo_dir_name) / "registry.db"),
+            project=project,
+            provider=test_repo_config.provider,
+            offline_store=offline_store,
+            online_store=online_store,
+            repo_path=repo_dir_name,
+        )
+        fs = FeatureStore(config=config)
+        environment = Environment(
+            name=project,
+            test_repo_config=test_repo_config,
+            feature_store=fs,
+            data_source=ds,
+            data_source_creator=offline_creator,
+        )
+
+        fvs = []
+        entities = []
+        try:
+            if create_and_apply:
+                entities.extend([driver(), customer()])
+                fvs.extend(
+                    [
+                        environment.driver_stats_feature_view(),
+                        environment.customer_feature_view(),
+                    ]
+                )
+                fs.apply(fvs + entities)
+
+            if materialize:
+                fs.materialize(environment.start_date, environment.end_date)
+
+            yield environment
+        finally:
+            offline_creator.teardown()
+            fs.teardown()
```

Comment:
Once data has been staged we should always be running teardown at the end. It doesn't seem like running teardown here is the right approach. Wouldn't we be leaving data staged if there is any test failure after [create_data_source](https://github.com/woop/feast-test-review/pull/1/files#diff-324627f3ffd597b6363cabcf3fcd67789eb39926622609040f9bdbfeef10550bR224)?

Pytest fixtures are typically used for setup/teardown. So to me it's a lot less abstraction/bad magic to do something like

```python
@pytest.fixture
def environment():
  e = Environment(config=TestRepoConfig())
  yield e
  e.teardown()

def my_e2e_test(e: Environment):
  # do stuff
```
and if we need to make the experience more magical it can be

```python
@pytest.fixture
def environment():
  e = construct_universal_test_environment(config options...)
  yield e
  e.teardown()

def my_e2e_test(e: Environment):
  # do stuff
```
but personally I still think it's going to be very brittle, so personally I would want to use something like the following. 

```python
@pytest.fixture
def environment():
  e = Environment(config=TestRepoConfig(config options...))
  e = stage_data(e)
  e = setup_sources(e)
  e = apply(e)
  e = materialize(e)
  yield e
  e.teardown()

def my_e2e_test(e: Environment):
  # do stuff
```
It's clearer as a user which operations are being performed in setting up the environment, and allows a test writer to create their own logic in a functional way to modify the environment. They can pick and choose which pre-built stages they want to use. In my mind this is a lot more stable of an approach than conditional logic. The conditional logic approach seems viable only as an eventual generalization when we have many test cases and we want to start merging logic, but it does seem a bit premature right now.